mackerel-plugin-flume/lib/flume.go

```diff
@@ -0,0 +1,220 @@
+			Metrics: []mp.Metrics{
+				{Name: "ConnectionCreatedCount", Label: "created", Diff: true},
+				{Name: "ConnectionClosedCount", Label: "closed", Diff: true},
+				{Name: "ConnectionFailedCount", Label: "failed", Diff: true},
+			},
+		},
+		"sink.event_drain_num.#": {
+			Label: labelPrefix + " Sink Event Drain Num",
+			Unit:  "integer",
+			Metrics: []mp.Metrics{
+				{Name: "EventDrainAttemptCount", Label: "attempt", Diff: true},
+				{Name: "EventDrainSuccessCount", Label: "success", Diff: true},
+			},
+		},
+		"source.append_num.#": {
+			Label: labelPrefix + " Source Append Num",
+			Unit:  "integer",
+			Metrics: []mp.Metrics{
+				{Name: "AppendAcceptedCount", Label: "accepted", Diff: true},
+				{Name: "AppendReceivedCount", Label: "received", Diff: true},
+			},
+		},
+		"source.append_batch_num.#": {
+			Label: labelPrefix + " Source Append Batch Num",
+			Unit:  "integer",
+			Metrics: []mp.Metrics{
+				{Name: "AppendBatchAcceptedCount", Label: "accepted", Diff: true},
+				{Name: "AppendBatchReceivedCount", Label: "received", Diff: true},
+			},
+		},
+		"source.event_num.#": {
+			Label: labelPrefix + " Source Event Num",
+			Unit:  "integer",
+			Metrics: []mp.Metrics{
+				{Name: "EventAcceptedCount", Label: "accepted", Diff: true},
+				{Name: "EventReceivedCount", Label: "received", Diff: true},
+			},
+		},
+		"source.connection.#": {
+			Label: labelPrefix + " Source Connection",
+			Unit:  "integer",
+			Metrics: []mp.Metrics{
+				{Name: "OpenConnectionCount", Label: "open"},
+			},
+		},
+	}
+}
+
+// FetchMetrics interface for mackerelplugin
+func (p *FlumePlugin) FetchMetrics() (map[string]float64, error) {
+	m, err := p.getMetrics()
+	if err != nil {
+		return nil, err
+	}
+	return p.parseMetrics(m), nil
+}
+
+func (p *FlumePlugin) getMetrics() (map[string]interface{}, error) {
+	res, err := http.Get(p.URI)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+	var m map[string]interface{}
+	dec := json.NewDecoder(res.Body)
+	if err := dec.Decode(&m); err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func (p *FlumePlugin) parseMetrics(metrics map[string]interface{}) map[string]float64 {
+	ret := make(map[string]float64)
+	for k, v := range metrics {
+		array := strings.Split(k, ".")
+		typeName := array[0]
+		componentName := array[1]
+		switch typeName {
+		case Channel:
+			p.parseChannel(ret, componentName, v.(map[string]interface{}))
+		case Sink:
+			p.parseSink(ret, componentName, v.(map[string]interface{}))
+		case Source:
+			p.parseSource(ret, componentName, v.(map[string]interface{}))
+		}
+	}
+
+	return ret
+}
+
+func (p *FlumePlugin) convertFloat64(value string) float64 {
+	f, _ := strconv.ParseFloat(value, 64)
+	return f
+}
+
+func (p *FlumePlugin) parseChannel(ret map[string]float64, componentName string, value map[string]interface{}) {
+	ret["channel.capacity." + componentName + ".ChannelCapacity"]             = p.convertFloat64(value["ChannelCapacity"].(string))
```

Comment:
Is flume componentName valid as mackerel metrics name?  I'm concerned that if flume can use `.` as component name, graph in mackerel may be broken.

Reply:
I think flume componentName valid as mackerel metrics name. because flume can not use . and = as component name.

Reply:
I see!